scarches/trainers/scpoli/trainer.py

```diff
@@ -0,0 +1,609 @@
+    n_workers: Integer
+        Passes the 'n_workers' parameter for the torch.utils.data.DataLoader class.
+    labeled_indices: list
+        List of integers that indicate the annotated data
+    pretraining_epochs: Integer
+        Number of epochs for pretraining
+    clustering: String
+        Clustering methodology to use, can be KMeans or Leiden.
+    clustering_res: Float
+        Clustering resolution to use for leiden clustering. Bigger values result in
+        finer clusters.
+    n_clusters: Integer
+        Number of clusters to set for KMeans algorithm.
+    unlabeled_weight: Float
+        Weight for loss computed including unlabeled samples
+    eta: Float
+        Weight for the landmark loss
+    seed: Integer
+        Define a specific random seed to get reproducable results.
+    """
+
+    def __init__(
+            self,
+            model,
+            adata,
+            labeled_indices: list = None,
+            pretraining_epochs: int = 0,
+            clustering: str = "leiden",
+            clustering_res: float = 1,
+            n_clusters: int = None,
+            unlabeled_weight: float = 0.001,
+            eta: float = 1,
+            **kwargs,
+    ):
+
+        super().__init__(model, adata, **kwargs)
+
+        self.eta = eta
+        self.clustering = clustering
+        self.n_clusters = n_clusters
+        self.unlabeled_weight = unlabeled_weight
+        self.clustering_res = clustering_res
+        self.pretraining_epochs = pretraining_epochs
+        self.use_early_stopping_orig = self.use_early_stopping
+
+        self.landmarks_labeled = None  # landmarks labeled cells (means)
+        self.landmarks_labeled_cov = None  # landmarks labeled cells (cov)
+        self.landmarks_unlabeled = None  # landmarks all cells (means)
+        self.best_landmarks_labeled = None  # cache for ES, to use best state
+        self.best_landmarks_labeled_cov = None  # cache for ES
+        self.best_landmarks_unlabeled = None  # cache for ES
+        self.landmark_optim = None  # landmark optimizer
+        # Set indices for labeled data
+        if labeled_indices is None:
+            self.labeled_indices = range(len(adata))
+        else:
+            self.labeled_indices = labeled_indices
+        self.update_labeled_indices(self.labeled_indices)
+
+        # Parse landmarks from model into right format
+        if self.model.landmarks_labeled["mean"] is not None:
+            self.landmarks_labeled = self.model.landmarks_labeled["mean"]
+            self.landmarks_labeled_cov = self.model.landmarks_labeled["cov"]
+        if self.landmarks_labeled is not None:
+            self.landmarks_labeled = self.landmarks_labeled.to(device=self.device)
+            self.landmarks_labeled_cov = self.landmarks_labeled_cov.to(
+                device=self.device
+            )
+
+    def train(self, n_epochs=400, lr=1e-3, eps=0.01):
+        begin_loader = time.time()
+        print("loaders init")
```

Comment:
@moinfar There are some lines here that are leftovers from some debugging i was doing

Reply:
Can you delete line 127, 129, 130, 131?

Reply:
also 126 is not needed

Reply:
Done.